tensorflow/compiler/mlir/hlo/lib/Analysis/userange_analysis.cc

```diff
@@ -0,0 +1,546 @@
+UserangeAnalysis::IntervalVector UserangeAnalysis::computeInterval(
+    Value value, const Liveness::OperationListT& operationList) {
+  assert(!operationList.empty() && "Operation list must not be empty");
+  size_t start = computeID(value, *operationList.begin());
+  size_t last = start;
+  UserangeAnalysis::IntervalVector intervals;
+  // Iterate over all operations in the operationList. If the gap between the
+  // respective operationIds is greater 1 create a new interval.
+  for (auto opIter = ++operationList.begin(), e = operationList.end();
+       opIter != e; ++opIter) {
+    size_t current = computeID(value, *opIter);
+    if (current - last > 2) {
+      intervals.emplace_back(UserangeAnalysis::UseInterval(start, last));
+      start = current;
+    }
+    last = current;
+  }
+  intervals.emplace_back(UserangeAnalysis::UseInterval(start, last));
+  return intervals;
+}
+
+/// Checks each operand inside the operation for its memory effects and
+/// separates them into read and write. Operands with read effects are added to
+/// the opToReadMap.
+void UserangeAnalysis::gatherMemoryEffects(Operation* op) {
+  if (OpTrait::hasElementwiseMappableTraits(op)) {
+    if (auto effectInterface = dyn_cast<MemoryEffectOpInterface>(op)) {
+      SmallPtrSet<Value, 2> readEffectSet;
+      SmallPtrSet<Value, 2> writeEffectSet;
+      for (auto operand : op->getOperands()) {
+        SmallVector<MemoryEffects::EffectInstance, 2> effects;
+        effectInterface.getEffectsOnValue(operand, effects);
+        for (auto effect : effects) {
+          if (isa<MemoryEffects::Write>(effect.getEffect())) {
+            writeEffectSet.insert(operand);
+          } else if (isa<MemoryEffects::Read>(effect.getEffect()))
+            readEffectSet.insert(operand);
+        }
+      }
+      opReadWriteMap.insert(
+          {op, std::make_pair(readEffectSet, writeEffectSet)});
+    }
+  }
+}
+
+/// Computes the ID for the operation. If the operation contains operands which
+/// have read effects, the returning ID will be odd. This allows us to
+/// perform a replace in place.
+size_t UserangeAnalysis::computeID(Value v, Operation* op) const {
+  size_t doubledID = operationIds.find(op)->second * 2;
+  auto mapIter = opReadWriteMap.find(op);
+  if (mapIter == opReadWriteMap.end())
+    return doubledID;
+  auto reads = mapIter->second.first;
+  auto writes = mapIter->second.second;
+  if (reads.contains(v) && !writes.contains(v))
+    return doubledID - 1;
+  return doubledID;
+}
+
+/// Merge two IntervalVectors into a new IntervalVector. Return a pair with the
+/// resulting IntervalVector and a boolean if there were interferences during
+/// merging.
+std::pair<UserangeAnalysis::IntervalVector, bool>
+UserangeAnalysis::intervalMerge(const IntervalVector& intervalA,
+                                const IntervalVector& intervalB) const {
+  IntervalVector mergeResult;
+  std::set_union(intervalA.begin(), intervalA.end(), intervalB.begin(),
```

Comment:
Would it not be "easier" to co-iterate the two vectors? It would also remove the need to erase elements from the vector.